File: leetcode/0094.Binary-Tree-Inorder-Traversal/Solution_test.go
package leetcode

import (
	"reflect"
	"testing"

	structures "github.com/nickmyb/algorithm-notes/structures/go"
)

func TestInorderTraversal(t *testing.T) {
	NULL := structures.NULL
	// 题目英文版给的 4 个示例，原样转录。
	// 注意中文版只有 3 个，缺了下面这棵大树——官方翻译滞后，以英文版为准。
	qs := []struct {
		name string
		in   []int
		want []int
	}{
		{"Example 1", []int{1, NULL, 2, 3}, []int{1, 3, 2}},
		{"Example 2", []int{1, 2, 3, 4, 5, NULL, 8, NULL, NULL, 6, 7, 9}, []int{4, 2, 6, 5, 7, 1, 3, 9, 8}},
		// 题解从 []int{} 起手而不是 var result []int，所以空树返回空切片而非 nil。
		// 这点要留意：LeetCode 判题对 nil 和空数组一视同仁，reflect.DeepEqual 不。
		{"Example 3 空树", []int{}, []int{}},
		{"Example 4 单节点", []int{1}, []int{1}},
	}

	for _, q := range qs {
		t.Run(q.name, func(t *testing.T) {
			got := inorderTraversal(structures.Ints2TreeNode(q.in))
			if !reflect.DeepEqual(got, q.want) {
				t.Fatalf("inorderTraversal(%v) = %v, want %v", structures.FormatInts(q.in), got, q.want)
			}
		})
	}
}

File: leetcode/0094.Binary-Tree-Inorder-Traversal/solution_test.py
R"""94. Binary Tree Inorder Traversal 的测试。

题解本体是 class Solution（和 LeetCode 模板一致），所以要先实例化。
"""

import pytest
from tree_node import build_tree

# 题目英文版给的 4 个示例，输入直接照抄题面里的层序数组。
# 中文版只有 3 个，缺了 Example 2 那棵大树——官方翻译滞后，以英文版为准。
#
# 每行第一列是用例名，它会成为 pytest 的用例 id（见文件末尾的 ids=），
# 这样 -v 输出里显示的是 [Example 1] 而不是默认的 [vals0-want0]。
CASES = [
    ("Example 1", [1, None, 2, 3], [1, 3, 2]),
    ("Example 2", [1, 2, 3, 4, 5, None, 8, None, None, 6, 7, 9], [4, 2, 6, 5, 7, 1, 3, 9, 8]),
    ("Example 3 空树", [], []),
    ("Example 4 单节点", [1], [1]),
]


@pytest.mark.parametrize(
    "name,vals,want", CASES, ids=[c[0] for c in CASES]
)
def test_inorder_traversal(solution, name, vals, want):
    got = solution.Solution().inorderTraversal(build_tree(vals))
    assert got == want, f"{name}: inorderTraversal({vals}) = {got}, want {want}"
